Add clickable menu

Qingwu-Li opened the pull request:
Hello fellow geeks,
I tring to implemented a built-in clickable menu feature,
Tested with wayland,

bynect commented:
hello. I like your idea and I also wanted to implement the buttons.

I tried your branch and noticed some problems with the text alignment

![img-1737672610](https://github.com/user-attachments/assets/3503e9d6-4085-404c-bdab-5c58768dd27f)

I think the best way would be to either specify a maximum amount of buttons allowed (then dmenu is used for extra action) or to wrap the buttons to a newline. squeezing all actions in a single line is impossible when there are a lot after all

Qingwu-Li committed:
Add built-in menu structure and functions

Implement menu structure and related functions,
preparing for clickable menu feature.
Add new menus array to preserve current actions.

Signed-off-by: LI Qingwu <Qing-wu.Li@leica-geosystems.com.cn>

Qingwu-Li committed:
Add settings for built-in clickable menu

Introduce built-in menu settings in dunstrc.
Add corresponding settings in settings.h.
Update settings_data.h with new menu-related options.
Preparing for built-in clickable menu.

Signed-off-by: LI Qingwu <Qing-wu.Li@leica-geosystems.com.cn>

Qingwu-Li committed:
Implement built-in clickable menu rendering

Add menu button drawing function.
Update layout calculations for menu height.
Integrate menu rendering into notification display.

Signed-off-by: LI Qingwu <Qing-wu.Li@leica-geosystems.com.cn>

Qingwu-Li committed:
Implement built-in menu button click handling

Add function to handle built-in menu clicks.
Trigger action invocation on menu button click.

Signed-off-by: LI Qingwu <Qing-wu.Li@leica-geosystems.com.cn>

Qingwu-Li committed:
Release resources for menu when invalidating actions

Add menu_free_array call in notification_invalidate_actions,
ensure proper cleanup of menu resources when invalidating actions

Signed-off-by: LI Qingwu <Qing-wu.Li@leica-geosystems.com.cn>

gujie-leica committed:
Improve revise settings parameter in built-in menu

- Revise fixed value with settings.width.max in built-in menu.
- Revise 'h_padding' parameter to align with the offical definition
ub built-in menu.

Signed-off-by: gujie <jie.gu@leica-geosystems.com.cn>

gujie-leica committed:
Add support round corner menu in built-in menu

Signed-off-by: gujie <jie.gu@leica-geosystems.com.cn>

gujie-leica commented:
We revised the settings parameter in the built-in menu and added support for a rounded corner menu in the built-in menu two commits ago.

bynect commented:
I will try the new changes 👍

gujie-leica commented:
Now we add key navigation inside of notification:

- left, right, up and down arrow keys select the button (action)
- enter key executes action
- add “focus“ when selected, the first key is the default focus key

bynect commented:
After this change

```
                        cairo_move_to(c, round(text_x * scale), round(text_y * scale));
```

It almost works correctly, with a minor offset to the text.

![img-1746477429](https://github.com/user-attachments/assets/47fe48b5-6d3c-439f-872d-202a98330915)

gujie-leica commented:
> 
> I think splitting the keyword part in a separate pr is better since this one already quite big.

Thank you for your reply. Basically, the drawing and navigation parts are separate, but they share some common changes in input.c. I will update it in the coming days.

gujie-leica committed:
Revise to use scale in draw built-in menu

Signed-off-by: gujie <jie.gu@leica-geosystems.com.cn>

bynect commented:
is this ready for review?

gujie-leica commented:
> is this ready for review?

Yes, I think so, I just rebase to resolve conflicts.  Sorry we're busy recently, didn't reponse in time.

bynect commented:
hello I finally tested it properly. codewise it seems ready to merge.
 I would like to add some considerations:
 - I'm not sure I would enable the menu by default since it is a new feature and it would change behavior
 - When the menu is active the old dmenu way of doing actions is still active. probably this is right, but we might consider disabling it when there are few actions
 - for consistency I would probably rename built_in_menu to enable_menu or button_menu

bynect commented:
also I think that the way the background can be improved. instead of having menu_frame_fill
something like `separator_color` can be used (which can be either `frame` or a color).
personally I would also add a way to change text color inside the menu. but that's something I can add myself.

bynect commented:
let me know what do you think

bynect commented:
<img width="341" height="163" alt="img-1756027602" src="https://github.com/user-attachments/assets/267894a0-eef6-4399-8525-cee15f316de7" />


long action names should be ellipsized or truncated to avoid this


<img width="341" height="227" alt="img-1756027874" src="https://github.com/user-attachments/assets/7bb441fb-0dca-4cf6-82af-71132bb2c08f" />

a big number of items is handled correctly 👍🏻 
but maybe a limit to the number of button should be added to keep in mind height limitation of the notification
so the excess buttons should be instead handled by dmenu

fwsmit commented on src/menu.c:
```diff
@@ -392,4 +392,90 @@ static gpointer context_menu_thread(gpointer data)
 
         return NULL;
 }
+
```
Right now the new code in menu.c is not integrated with the old code. This could lead to confusion with other similar functions in this file. I would suggest to either
1. create a separate dmenu.c for all dmenu stuff and renaming those functions to reflect the fact that there are multiple kinds of menu's
or 2.  Merge some of the new functions with old ones and rename the other existing functions to avoid confusion.

fwsmit reviewed (commented):
I have some suggestions that could improve the code, avoiding issues in the future. Other than that it looks good

fwsmit commented:
Also, could you add some tests, for example for `calculate_menu_height`

ClairDePrince commented:
Any update?

fedang commented:
a few touch-ups to the layout calculation and a few tests and this should be mergeable
